padertorch/contrib/examples/speaker_classification/supervised/evaluate.py

```diff
@@ -0,0 +1,166 @@
+"""
+Example calls:
+
+On single CPU (slow):
+python -m padertorch.contrib.examples.speaker_classification.supervised.evaluate with model_path=<path/to/trained/model>
+
+On GPU:
+python -m padertorch.contrib.examples.speaker_classification.supervised.evaluate with model_path=<path/to/trained/model> device=0 batch_size=16
+
+On multiple CPUs:
+mpiexec -np 8 python -m padertorch.contrib.examples.speaker_classification.supervised.evaluate with model_path=<path/to/trained/model>
+"""
+import os
+from pathlib import Path
+import warnings
+
+from sacred import Experiment, commands
+import numpy as np
+import torch
+
+from dlp_mpi import COMM, IS_MASTER, MASTER, split_managed
+
+from paderbox.io.new_subdir import get_new_subdir
+from paderbox.io import load_json, dump_json
+
+import padertorch as pt
+
+from .data import get_datasets
+
+ex = Experiment('speaker_clf')
+
+
+@ex.config
+def defaults():
+    model_path = None
+    assert model_path is not None, (
+        'model_path cannot be None.\n'
+        'Start the evaluation with "python -m padertorch.contrib.examples.'
+        'speaker_classification.evaluate with model_path=<path/to/trained/model>"'
+    )
+    load_ckpt = 'ckpt_best_loss.pth'
+    batch_size = 1
+    device = 'cpu'
+    store_misclassified = True
+
+
+@ex.automain
+def main(_run, model_path, load_ckpt, batch_size, device, store_misclassified):
+    if IS_MASTER:
+        commands.print_config(_run)
+
+    model_path = Path(model_path)
+    eval_dir = get_new_subdir(
+        model_path / 'eval', id_naming='time', consider_mpi=True
+    )
+    # perform evaluation on a sub-set (10%) of the dataset used for training
+    config = load_json(model_path / 'config.json')
+    database_json = config['database_json']
+    dataset = config['dataset']
+
+    model = pt.Model.from_config(config['model'])
+    model = model.load_checkpoint(
+        model_path / 'checkpoints' / load_ckpt, consider_mpi=True
+    )
+    model.to(device)
+    # Turn on evaluation mode for, e.g., BatchNorm and Dropout modules
+    model.eval()
+
+    _, _, test_set = get_datasets(
+        model_path, database_json, dataset, batch_size,
+        prefetch=device != 'cpu'
+    )
+    with torch.no_grad():
+        summary = dict(
+            misclassified_examples=dict(),
+            correct_classified_examples=dict(),
+            hits=list()
+        )
+        for batch in split_managed(
+            test_set, is_indexable=batch_size == 1 and device == 'cpu',
```

Comment:
Be careful with `is_indexable=batch_size == 1`. If `is_indexable` is `False`, and you use MPI, then each process iterates over the dataset. This is expensive, because you do all the data loading in the dataset.

Reply:
Thanks for the hint. Now, the dataset is always indexable if evaluation is performed on the CPU (https://github.com/fgnt/padertorch/pull/57/commits/5517ed73f58728726506cfafd225333a8aed08c7).